types/preview/@ember/template-compilation/index.d.ts

```diff
@@ -0,0 +1,26 @@
+/**
+ * References:
+ *  - https://github.com/emberjs/rfcs/pull/496
+ *  - https://github.com/emberjs/rfcs/pull/779
+ */
+declare module '@ember/template-compilation' {
+  interface TemplateFactory {
+    __htmlbars_inline_precompile_template_factory: any;
+  }
+
+  type PrecompileTemplateOptions =
+    | {
+        strictMode?: false;
+        moduleName?: string;
+      }
+    | {
+        strictMode: true;
+        moduleName?: string;
+        scope: () => Record<string, unknown>;
+      };
```

Comment:
I have a mild preference for this layout, which is clearer about the shared code and what exactly the options are:

```suggestion
interface MaybeModuleName {
  moduleName?: string;
}

interface LooseMode extends MaybeModuleName {
  strictMode?: false;
}

interface StrictMode extends MaybeModuleName {
  strictMode: true;
  scope: () => Record<string, unknown>;
}

type PrecompileTemplateOptions = LooseMode | StrictMode;
```

Reply:
applied!